Added support for ROS2 Jazzy and Ubuntu 24.04

nathanshankar opened the pull request:
- Addressed segmentation faults caused by the latest Ceres version on Ubuntu 24.04
- Added support for ROS2 Jazzy by fixing header includes and adjustments in the ros2visualizer module
- config files D455 Stereo Mode: Using stereo IR streams + IMU.
- config files D455 Mono Mode: Using mono RGB + IMU.
![image](https://github.com/user-attachments/assets/4ebb3d12-3797-4657-aa56-bac9157e61e1)

nathanshankar commented:
> I can confirm I can build/run on ubuntu24/jazzy with these changes.
> 
> But for some reason with same sensor (Luxonis OAK-Pro-W) and same open_vins settings the initialisation does not work:
> 
> ```
> [run_subscribe_msckf-6] [init]: disparity is 0.069,0.045 (2.00 thresh)
> [run_subscribe_msckf-6] [init]: failed static init: no accel jerk detected
> [run_subscribe_msckf-6] [TIME]: 0.0048 seconds total (208.4 hz, 43.98 ms behind)
> ```
> 
> did anyone experience this?

Do you have a rosbag, so I can reproduce this?

asimonov commented:
I traced the problem to change in Luxonis Oak-d-pro driver for ROS Jazzy: https://github.com/luxonis/depthai-ros/pull/696/files

Without new parameters IMU was running at 0.2 Hz and that was an issue for OpenVINS

asimonov commented:
this PR works on Jazzy, BTW. I have tested

goldbattle commented:
Is it possible to make this backwards compatible? I know the older versions can't have these header changes.
For the ceres fix this PR has a compatible method so the older versions can still be supported. https://github.com/rpng/open_vins/pull/520

nathanshankar committed:
modified:   ov_core/CMakeLists.txt
	modified:   ov_core/cmake/ROS1.cmake
	modified:   ov_core/cmake/ROS2.cmake
	modified:   ov_eval/CMakeLists.txt
	modified:   ov_eval/cmake/ROS1.cmake
	modified:   ov_eval/cmake/ROS2.cmake
	modified:   ov_init/CMakeLists.txt
	modified:   ov_init/cmake/ROS1.cmake
	modified:   ov_init/cmake/ROS2.cmake
	modified:   ov_msckf/CMakeLists.txt
	modified:   ov_msckf/cmake/ROS1.cmake
	modified:   ov_msckf/cmake/ROS2.cmake
	modified:   ov_msckf/src/ros/ROS2Visualizer.h
	modified:   ov_msckf/src/ros/ROSVisualizerHelper.h
	modified:   ov_msckf/src/update/UpdaterHelper.cpp

nathanshankar committed:
modified:   .github/workflows/build_ros2.yml
	new file:   Dockerfile_ros2_24_04

nathanshankar committed:
modified:   ov_core/CMakeLists.txt
	modified:   ov_core/cmake/ROS1.cmake
	modified:   ov_core/cmake/ROS2.cmake
	modified:   ov_eval/CMakeLists.txt
	modified:   ov_eval/cmake/ROS1.cmake
	modified:   ov_eval/cmake/ROS2.cmake
	modified:   ov_init/CMakeLists.txt
	modified:   ov_init/cmake/ROS1.cmake
	modified:   ov_init/cmake/ROS2.cmake
	modified:   ov_msckf/CMakeLists.txt
	modified:   ov_msckf/cmake/ROS1.cmake
	modified:   ov_msckf/cmake/ROS2.cmake

nathanshankar committed:
modified:   ov_msckf/src/ros/ROS2Visualizer.cpp
	modified:   ov_msckf/src/ros/ROS2Visualizer.h
	modified:   ov_msckf/src/run_simulation.cpp
	modified:   ov_msckf/src/run_subscribe_msckf.cpp

nathanshankar reviewed (commented):
@goldbattle  the latest changes are backward compatible and support the header changes as requested :) There were some issues in the jazzy build for object destruction and those are addressed too

notokay00 commented:
@nathanshankar @asimonov 
My university team and I are building an autonomous UAV, and I could use a check on our OS/ROS setup for our companion computer.
Right now, we have our SITL simulation (Gazebo Harmonic and ArduPilot) running perfectly on a laptop using Ubuntu 22.04 and ROS 2 Humble. We are thinking of using OpenVINS and OpenCV, pulling video via the ros_gz_bridge.
Here is the roadblock: our actual flight hardware is a Raspberry Pi 5 , which do not support 22.04 natively . (could try docker as a last resort though). 
We are planning to migrate our physical hardware stack to Ubuntu 24.04 (Noble) and ROS 2 Jazzy to natively support the Pi 5. The proposed stack for the Pi 5 companion computer:
Ubuntu 24.04 Desktop (64-bit)
ROS 2 Jazzy
OpenVINS (built for Jazzy)
micro_ROS_agent

My questions:
Has anyone successfully run this VIO stack on Jazzy with a Pi 5?
Does the pipeline translate directly from Humble to Jazzy, or are there major breaking changes or dependency issues I should prepare the team for?

Any advice or inaccuracy would be hugely appreciated before we fully commit to this architecture. Thanks!

Roboticistprogrammer commented:
Hi, i am looking for this setup too. is there update on when it might be merged? @goldbattle

nathanshankar commented:
@notokay00 I worked on this a while back https://github.com/nathanshankar/rpi-bullseye-ros2. It should work with my jazzy openvins version https://github.com/nathanshankar/open_vins. However the build time might take a while. I mostly work with AGVs so I am not sure if ardupilot sitl is supported for jazzy, but you can try with docker.

notokay00 commented:
@nathanshankar Thanks for your reply ,  your repo worked with jazzy , was able to get stable result , and yes ardupilot does support jazzy .
